### ISSUE TITLE
Better handling of splatted keyword arguments and with_collection

### ISSUE THREAD

**Issue by daniel-rikowski** [ViewComponent/view_component#2020]
## Feature request

Allow components to have unnamed keyword arguments when using `with_collection`

### Motivation

This would allow using generic initialize methods with unnamed parameters and provide more flexible opportunities for composition, especially in regards of DRYing up initializers.

Also, I can imagine, this would become a necessity with #2004 

### Details

Currently these scenarious are not possible:

#### Example 1: (concept from [ViewComponentContrib](https://github.com/palkan/view_component-contrib))

```ruby
class ApplicationViewComponent < ViewComponent::Base
  extend Dry::Initializer[undefined: false]
end

class ButtonComponent < ApplicationViewComponent
  option :button

  erb_template <<~ERB
    <%= button.inspect %>
  ERB
end
```

NB: `extend Dry::Initializer` adds this method definition:

```
def initialize(...)
  __dry_initializer_initialize__(...)
end
```

(`...` is verbatim Ruby code)

#### Example 2: (meta programming)

```ruby
module InitHelpers
  extend ActiveSupport::Concern

  class_methods do
    def record_name = name.demodulize.underscore.chomp("_component").to_sym
  end

  included do
    attr_reader self.record_name
  end

  def init_component(**options)
    attr = self.class.record_name
    instance_variable_set(:"@#{attr}", options[attr])
  end
end

class AvatarComponent < ViewComponent::Base
  include InitHelpers

  def initialize(**)
    init_component(**)
  end

  erb_template <<~ERB
    <%= avatar.inspect %>
  ERB
end
```

There is an explicit check to exclude unnamed keyword arguments: 

https://github.com/ViewComponent/view_component/blob/66dcb7cc9956a52cbbd7fb12de98b90affd756e4/lib/view_component/base.rb#L695-L698

I did some rummaging and found this commit https://github.com/ViewComponent/view_component/commit/56a6560160795b5941ff67edcb56ea0878ae17fa It mentions that Ruby 3.1 `def m(*)` is implicitly the same as `def m(*,**)` 

If that's the case it does make sense to forbid unnamed splatted kwargs, but I assume this check can be relaxed for other Ruby versions.

### PR

I'm happy to make a PR, but first I want to discuss the proper way to do this.

**Comment by reeganviljoen** [comment]
@daniel-rikowski would you be able to add some failing tests that we could use as a spec

**Comment by daniel-rikowski** [comment]
@reeganviljoen How shall I approach this? Do you want me to create a PR with just the failing tests? Or shall I post them here before?

Anyway, I think I found the root cause.

If you have a component without an `initialize` method, `instance_method(:initialize)` in this line

https://github.com/ViewComponent/view_component/blob/66dcb7cc9956a52cbbd7fb12de98b90affd756e4/lib/view_component/base.rb#L708-L710

returns `ActionDispatch::Routing::UrlFor#initialize(...)`. 😮  

This stems from the `ActionView::Base` parent class of all view components. (I have no idea why this doesn't return `ViewComponent::Base#initialize` or `ActionView::Base#initialize`.)

In other words: Without an overridden initialize method, the `splatted_keyword_argument_present?` method actually checks `ActionDispatch::Routing::UrlFor#initialize`. (So no Ruby bug 😌 😄)

The following change does seem to fix the whole problem (still, some tests need to be adapted)

```ruby
      def initialize_parameters
        # Instead of:
        # @initialize_parameters ||= instance_method(:initialize).parameters
        # This:
        @initialize_parameters ||=
            instance_method(:initialize).then { |m| m.owner < ViewComponent::Base ? m.parameters : [] }
      end
```

**Comment by reeganviljoen** [comment]
@daniel-rikowski i have been quite busy lately but I will see if I can find some time to add failing tests and then your research should really come in handy

**Comment by daniel-rikowski** [comment]
I invested a few hours over the weekend. Although the failing tests and the fix are fairly simple, I can't figure out when or why `instance_method(:initialize)` returns `ActionDispatch::Routing::UrlFor#initialize(...)` instead of `ViewComponent::Base#initialize(*)`.

Here's are failing tests:

```ruby
# frozen_string_literal: true

require "test_helper"

module ViewComponent
  class UnnamedArgumentsTest < TestCase
    class DynamicComponentBase < ViewComponent::Base
      def setup_component(**attributes)
        # This method is somewhat contrived, it's intended to mimic features available in the dry-initializer gem.
        model_name = self.class.name.demodulize.delete_suffix('Component').underscore.to_sym
        instance_variable_set(:"@#{model_name}", attributes[model_name])
        define_singleton_method(model_name) { instance_variable_get(:"@#{model_name}") }
      end
    end

    class OrderComponent < DynamicComponentBase
      def initialize(**)
        setup_component(**)
      end
      def call
        "<div data-name='#{order.number}'><h1>#{order.number}</h1></div>".html_safe
      end
    end

    class CustomerComponent < DynamicComponentBase
      def initialize(...)
        setup_component(...)
      end
      def call
        "<div data-name='#{customer.name}'><h1>#{customer.name}</h1></div>".html_safe
      end
    end

    def setup
      @customers = [OpenStruct.new(name: "Taylor"), OpenStruct.new(name: "Rowan")]
      @orders = [OpenStruct.new(name: "O-2024-0004"), OpenStruct.new(name: "B-2024-0714")]
    end

    def test_supports_components_with_argument_forwarding
      render_inline(CustomerComponent.with_collection(@customers))
      assert_selector("*[data-name='#{@customers.first.name}']", text: @customers.first.name)
      assert_selector("*[data-name='#{@customers.last.name}']", text: @customers.last.name)
    end

    def test_supports_components_with_unnamed_splatted_arguments
      render_inline(OrderComponent.with_collection(@orders))
      assert_selector("*[data-name='#{@orders.first.number}']", text: @orders.first.number)
      assert_selector("*[data-name='#{@orders.last.number}']", text: @orders.last.number)
    end
  end
end
```

and this is the fix:

```ruby
module ViewComponent
  class Base < ActionView::Base
    class << self
      def splatted_keyword_argument_present?
        # initialize_parameters.flatten.include?(:keyrest) &&
        #   !initialize_parameters.include?([:keyrest, :**]) # Un-named splatted keyword args don't count!
        initialize_parameters.flatten.include?(:keyrest)
      end
    end
  end
end
```

**BUT:**

Now `RenderingTest#test_collection_component_missing_default_parameter_name` fails.

That component has no own `initialize` method:

```ruby
class MissingDefaultCollectionParameterComponent < ViewComponent::Base
  def call
  end
end
```
So it doesn't (=shouldn't) accept a collection argument and the test verifies, that ViewComponent detects this. 

But in this case it fails, because `instance_method(:initialize)` returns `ActionDispatch::Routing::UrlFor#initialize(...)` which _does_ accept it.

In other words:

`instance_method(:initialize)` returns `#<UnboundMethod: ActionDispatch::Routing::UrlFor#initialize(...) >`
`method(:initialize)` is `#<Method: #<Class:MissingDefaultCollectionParameterComponent>(Class)#initialize(*)>`


I have no idea why this happens, or how to circumvent it. My original idea to check `instance_method(:initialize).owner < ViewComponent::Base` would work, but that would also disable any `initialize` method coming from a module, e.g. `Dry::Initializer`.
 
I tried to reproduce this with minimal code, but I was unable to define a class in which `instance_method(:initialize)` and `method(:initialize)` pointed to different methods. I guess my Ruby knowledge is just limited here.

**Comment by MyklClason** [comment]
Thanks for referring to my abstract components issue. Reviewing what you have here, I'm not sure what you are suggesting is ideal. There is some merit in avoiding duplication via the `DynamicComponentBase` class, but there's actually a good reason why I suggest creating the base components. This is because the base component works really good as a place to put helper methods and other things that may be used across components that use the same base component. Concerns are a better solution for helpers not limited to a specific kind of input.

Indeed, rather than adding a dynamic base class. It is probably more helpful to modify the generators or make it easier to create your own custom generators that make use of dry-effects if you want. 

Here is how I would implement the `OrderComponent` and `CustomerComponent` you had. I used two different ways of handling "data_key" and "data_value". The first moves it to the base class, and the other just handles the key and value inline.

```
class DesignSystem::H1Component < Record::BaseComponent
  attr_reader :key, :value
  
  def initializer(key:, value:)
    @key = key
    @value = value
  end
  
  def call
    "<div data-#{key}='#{value}'><h1>#{value}</h1></div>".html_safe
  end
end

class Order::BaseComponent < ApplicationComponent
  attr_reader :order
  
  def initialize(order:)
    @order = order
  end

  def data_key
    :number
  end
  
  def data_value
    order.number
  end
end

class Order::H1Component < Order::BaseComponent
  def call
    DesignSystem::H1Component.new(key: data_key, value: data_value)
  end
end

class Customer::BaseComponent < ApplicationComponent
  attr_reader :customer
  
  def initialize(customer:)
    @customer = customer
  end
end

class Customer::H1Component < Customer::BaseComponent
  def call
    DesignSystem::H1Component.new(key: :name, value: customer.name)
  end
end
```

However, to solve the original issue, I did have ChatGPT assist with quickly generating a solution that uses class_eval. Though I used it for the example above, so it will work with things like: `Customer::TableComponent.new(customer:)` and `Customer::Invoice::TableComponent.new(customer_invoice:)`. It also works with `Customer::Invoice::TableComponent.new()`  which is the same as `Customer::Invoice::TableComponent.new(customer_invoice: nil)`

```
class DynamicComponentBase < ViewComponent::Base
  def self.inherited(subclass)
    super
    subclass.class_eval do
      parts = name.split("::").map(&:underscore)
      parts = parts[0..-2] if parts.size > 1
      component_name = parts.join("_")
      
      attr_reader component_name

      define_method(:initialize) do |**args|
        instance_variable_set("@#{component_name}", args[component_name.to_sym])
      end
    end
  end
end
```

Might actually use it myself to dry up the places where an initializer isn't need.

**Comment by daniel-rikowski** [comment]
@MyklClason Now I feel bad, because you put so much work into your answer 😩  

The code from the test was actually just intended to mimic some of the functionality of `dry-initializer` without introducing that gem as a new dependency into the test suite. I'm not looking for new ways to implement these features.

In a nutshell, I want to have this, more or less: https://github.com/palkan/view_component-contrib#hanging-initialize-out-to-dry

I.e. do this...

```ruby
class ApplicationViewComponent
  extend Dry::Initializer
end

class CustomerComponent < ApplicationViewComponent
  option :customer
end
```

... and be done with it.

But ViewComponent **actively** prevents these components to be used in a collection. 

(They need to have an explicit `initialize` method with named arguments, while a major purpose of `dry-initializer` is getting rid of said methods.)

**Comment by daniel-rikowski** [comment]
Anyway, even if it is decided, that my use case is not worth the trouble, it is still a possibility, that some day the Rails developers change the signature of the method, which made this whole workaround necessary.

E.g. they change `ActionView::Routing:UrlFor#initialize` from this

```ruby
def initialize(...)
  @_routes = nil
  super
end
```

to - let's say - this:

```ruby
def initialize(*, **options)
  @_routes = nil
  @_options = options
  super
end
```

(Unlikely, admittedly, but it is a possibility.)

Then, the afformentioned parameters check would no longer help in preventing misconfigured components, because then, many components get their _named_ splatted arguments from `ActionView::Routing:UrlFor#initialize` (Without their developers ever knowing...)

IMHO this is a much deeper problem than just preventing certain component class definitions.

**Comment by MyklClason** [comment]
Lol, no problem. My initializers are more DRY now lol. So I'm happy enough.

However, isn't isn't that what the last edit did?

```
class ApplicationViewComponent
  include BaseComponentInitializer # extend Dry::Initializer
end

class CustomerComponent < ApplicationViewComponent
  base_initializer :customer # option :customer
end
```

Of course, you can change the method name so it uses "option" instead of "base_initializer".

I have seen the "_routes" you mention, but I just ignore it.

Here is a version that in theory mimics what you( had except it uses "include" rather than "extend". Been a while since I looked into the differences of those.

```
module Dry::Initializer
  extend ActiveSupport::Concern

  included do
    def self.option(*attributes)
      class_eval do
        attr_reader *attributes

        define_method(:initialize) do |**kargs|
          attributes.each do |attribute|
            instance_variable_set("@#{attribute}", kargs[attribute.to_sym])
          end
        end
      end
    end
  end
end

class ApplicationViewComponent
  extend Dry::Initializer
end

class CustomerComponent < ApplicationViewComponent
  option :customer
end
```

**Comment by joelhawksley** [comment]
Closed by https://github.com/ViewComponent/view_component/pull/2332. Mind giving the `v4` branch a try to see if it resolves your concerns?
